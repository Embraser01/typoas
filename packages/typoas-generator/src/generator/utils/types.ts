import { ReferenceObject, SchemaObject } from 'openapi3-ts';
import { factory, TypeNode, SyntaxKind } from 'typescript';
import { Context } from '../../context';
import {
  hasUnsupportedIdentifierChar,
  sanitizeTypeIdentifier,
} from './operation-name';

const ENUM_SUPPORTED_TYPE = ['integer', 'number', 'string', 'null', 'boolean'];

export function createTypeFromSchema(
  schemaOrRef: SchemaObject | ReferenceObject | undefined,
  ctx: Context,
): TypeNode {
  let node: TypeNode = factory.createKeywordTypeNode(SyntaxKind.AnyKeyword);

  if (!schemaOrRef) {
    return node;
  }
  if (schemaOrRef.$ref) {
    const ref = ctx.resolveReference('schemas', schemaOrRef.$ref);
    if (!ref) {
      throw new Error(`$ref '${schemaOrRef.$ref}' wasn't found`);
    }
    node = factory.createTypeReferenceNode(
      factory.createIdentifier(sanitizeTypeIdentifier(ref.name)),
    );
  }

  const schema: SchemaObject = schemaOrRef;
  if (schema.enum && schema.type && ENUM_SUPPORTED_TYPE.includes(schema.type)) {
    node = factory.createUnionTypeNode(
      schema.enum
        .map((e) => {
          if (e === null) return factory.createNull();
          if (typeof e === 'boolean') {
            return e ? factory.createTrue() : factory.createFalse();
          }
          if (typeof e === 'number') {
            return factory.createNumericLiteral(e);
          }
          return factory.createStringLiteral(e, true);
        })
        .map((e) => factory.createLiteralTypeNode(e)),
    );
  } else if (schema.oneOf && schema.oneOf.length) {
    node = factory.createUnionTypeNode(
      schema.oneOf.map((s) => createTypeFromSchema(s, ctx)),
    );
  } else if (schema.anyOf && schema.anyOf.length) {
    node = factory.createUnionTypeNode(
      schema.anyOf.map((s) => createTypeFromSchema(s, ctx)),
    );
  } else {
    switch (schema.type) {
      case 'integer':
      case 'number':
        node = factory.createKeywordTypeNode(SyntaxKind.NumberKeyword);
        break;
      case 'string':
        switch (schema.format) {
          case 'date':
          case 'date-time':
            node = factory.createTypeReferenceNode(
              factory.createIdentifier('Date'),
            );
            break;
          default:
            node = factory.createKeywordTypeNode(SyntaxKind.StringKeyword);
            break;
        }
        break;
      case 'boolean':
        node = factory.createKeywordTypeNode(SyntaxKind.BooleanKeyword);
        break;
      case 'object': {
        const hasProperties =
          schema.properties && Object.keys(schema.properties).length;

        if (hasProperties) {
          node = factory.createTypeLiteralNode(
            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
            Object.entries(schema.properties!).map(([key, val]) =>
              factory.createPropertySignature(
                undefined,
                hasUnsupportedIdentifierChar(key)
                  ? factory.createStringLiteral(key, true)
                  : factory.createIdentifier(key),
                schema.required?.includes(key)
                  ? undefined
                  : factory.createToken(SyntaxKind.QuestionToken),
                createTypeFromSchema(val, ctx),
              ),
            ),
          );
        }
        if (schema.additionalProperties) {
          const additionalPropsNode = factory.createTypeLiteralNode([
            factory.createIndexSignature(
              undefined,
              undefined,
              [
                factory.createParameterDeclaration(
                  undefined,
                  undefined,
                  undefined,
                  factory.createIdentifier('key'),
                  undefined,
                  factory.createKeywordTypeNode(SyntaxKind.StringKeyword),
                  undefined,
                ),
              ],
              schema.additionalProperties === true
                ? factory.createKeywordTypeNode(SyntaxKind.AnyKeyword)
                : createTypeFromSchema(schema.additionalProperties, ctx),
            ),
          ]);

          node = hasProperties
            ? factory.createIntersectionTypeNode([node, additionalPropsNode])
            : additionalPropsNode;
        }
        break;
      }
      case 'null':
        node = factory.createLiteralTypeNode(factory.createNull());
        break;
      case 'array':
        node = factory.createArrayTypeNode(
          createTypeFromSchema(schema.items, ctx),
        );
        break;
    }
  }

  if (schema.nullable) {
    node = factory.createUnionTypeNode([
      node,
      factory.createLiteralTypeNode(factory.createNull()),
    ]);
  }
  return node;
}
