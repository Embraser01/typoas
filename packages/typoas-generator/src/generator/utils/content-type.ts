import {
  ContentObject,
  ReferenceObject,
  SchemaObject,
} from 'openapi3-ts/oas31';
import { Context } from '../../context';

// This key is a simple hack for the createTypeFromSchema function
// for it to generate a Blob type that does not exist by default in json schema.
export const TYPOAS_BLOB_TYPE_KEY = 'x-typoas-blob-type';

export function getContentTypeSchema(
  content: ContentObject | undefined,
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  ctx: Context,
): ReferenceObject | SchemaObject | undefined {
  if (!content) {
    return undefined;
  }
  if (content['application/json']) {
    return content['application/json'].schema;
  }

  if (content['text/plain']) {
    return { type: 'string' };
  }

  if (content['application/octet-stream']) {
    return { [TYPOAS_BLOB_TYPE_KEY]: true };
  }
}
