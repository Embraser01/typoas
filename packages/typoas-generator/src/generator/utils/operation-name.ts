function lowerCase(str: string) {
  return str.toLowerCase();
}

function upperCase(str: string) {
  return str.toUpperCase();
}

/**
 * Convert string to camelCase text.
 */
function camelCase(str: string): string {
  return str
    .replace(/[-./]/g, ' ') //convert all hyphens to spaces
    .replace(/\s[a-z]/g, upperCase) //convert first char of each word to UPPERCASE
    .replace(/\s+/g, '') //remove spaces
    .replace(/^[A-Z]/g, lowerCase); //convert first char to lowercase
}

function pascalCase(str: string): string {
  return camelCase(str).replace(/^[a-z]/, upperCase);
}

export function hasUnsupportedIdentifierChar(key: string): boolean {
  return /[-/.+@]/.test(key);
}

export function sanitizeOperationIdName(op: string): string {
  return camelCase(op);
}

export function sanitizeTypeIdentifier(type: string): string {
  return pascalCase(type);
}
