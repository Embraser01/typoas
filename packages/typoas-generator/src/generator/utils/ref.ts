import { Expression, factory, TypeNode } from 'typescript';

export const IMPORT_RUNTIME = 'r';

export enum ExportedRef {
  HttpMethod = 'HttpMethod',
  BaseServerConfiguration = 'BaseServerConfiguration',
  HttpLibrary = 'HttpLibrary',
  RefResolver = 'RefResolver',
  IsomorphicFetchHttpLibrary = 'IsomorphicFetchHttpLibrary',
  SchemaRefResolver = 'SchemaRefResolver',
  ApiException = 'ApiException',
  AuthProvider = 'AuthProvider',
  BasicAuthConfig = 'BasicAuthConfig',
  BearerAuthConfig = 'BearerAuthConfig',
  BaseFlowConfig = 'BaseFlowConfig',
  SecurityAuthentication = 'SecurityAuthentication',
  ApiKeySecurityAuthentication = 'ApiKeySecurityAuthentication',
  HttpSecurityAuthentication = 'HttpSecurityAuthentication',
  OAuth2SecurityAuthentication = 'OAuth2SecurityAuthentication',
  applyTemplating = 'applyTemplating',
  applyTransforms = 'applyTransforms',
  isCodeInRange = 'isCodeInRange',
  serializeParameter = 'serializeParameter',
}

export function createRuntimeRefProperty(exportedRef: ExportedRef): Expression {
  return factory.createPropertyAccessExpression(
    factory.createIdentifier(IMPORT_RUNTIME),
    exportedRef,
  );
}

export function createRuntimeRefType(
  exportedRef: ExportedRef,
  typeArguments?: readonly TypeNode[],
): TypeNode {
  return factory.createTypeReferenceNode(
    factory.createQualifiedName(
      factory.createIdentifier(IMPORT_RUNTIME),
      factory.createIdentifier(exportedRef),
    ),
    typeArguments,
  );
}

export function createJSONParseWrapper(data: unknown): Expression {
  return factory.createCallExpression(
    factory.createPropertyAccessExpression(
      factory.createIdentifier('JSON'),
      'parse',
    ),
    undefined,
    [factory.createStringLiteral(JSON.stringify(data), true)],
  );
}
