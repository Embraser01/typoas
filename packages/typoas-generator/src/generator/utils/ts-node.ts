import {
  createPrinter,
  EmitHint,
  factory,
  NewLineKind,
  Node,
  NodeFlags,
  SyntaxKind,
} from 'typescript';

export const getStringFromNode = (node: Node): string => {
  const printer = createPrinter({ newLine: NewLineKind.LineFeed });
  const resultFile = factory.createSourceFile(
    [],
    factory.createToken(SyntaxKind.EndOfFileToken),
    NodeFlags.Const,
  );
  return printer.printNode(EmitHint.Unspecified, node, resultFile);
};
