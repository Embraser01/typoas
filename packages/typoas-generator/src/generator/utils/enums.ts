import { SchemaObject } from 'openapi3-ts';
import { factory, EnumMember } from 'typescript';
import { screamingSnakeCase } from './operation-name';

export function canConvertSchemaToEnum(schema: SchemaObject): boolean {
  return (
    schema.type === 'string' &&
    schema.enum !== undefined &&
    schema.enum.every((e) => ![null, '', undefined].includes(e)) &&
    !schema.nullable
  );
}

export function createEnumMembersFromSchema(
  schema: SchemaObject,
): EnumMember[] {
  if (canConvertSchemaToEnum(schema)) {
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    return schema.enum!.map((e) => {
      return factory.createEnumMember(
        factory.createIdentifier(screamingSnakeCase(e)),
        factory.createStringLiteral(e, true),
      );
    });
  }
  throw new Error('Cannot create enum members from this schema');
}
