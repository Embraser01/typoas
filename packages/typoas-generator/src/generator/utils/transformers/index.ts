import {
  ArrayLiteralExpression,
  Expression,
  factory,
  ObjectLiteralElementLike,
  ObjectLiteralExpression,
} from 'typescript';
import { ReferenceObject, SchemaObject } from 'openapi3-ts';
import { Context } from '../../../context';
import { isTransformableLeafDate } from './date';
import { findLeafTransforms, TransformField } from './leaf-transformer-base';
import { createRuntimeRefProperty, ExportedRef } from '../ref';

export function createSchemaTransforms(
  schema: SchemaObject | ReferenceObject,
  ctx: Context,
): ObjectLiteralExpression | null {
  const transforms: ObjectLiteralElementLike[] = [];

  // Date transform
  const rawTransforms = findLeafTransforms(schema, isTransformableLeafDate);
  if (rawTransforms.length) {
    transforms.push(
      factory.createPropertyAssignment(
        'date',
        createFromRawTransforms(rawTransforms, ctx),
      ),
    );
  }

  if (!transforms.length) {
    return null;
  }
  return factory.createObjectLiteralExpression(transforms);
}

function createFromRawTransforms(
  rawTransforms: TransformField[],
  ctx: Context,
): ArrayLiteralExpression {
  return factory.createArrayLiteralExpression(
    rawTransforms.map((transformFields) =>
      factory.createArrayLiteralExpression(
        transformFields.map(([type, data]) => {
          const items: Expression[] = [
            factory.createPropertyAccessExpression(
              createRuntimeRefProperty(ExportedRef.TransformType),
              factory.createIdentifier(type.toUpperCase()),
            ),
          ];

          if (typeof data === 'string') {
            items.push(factory.createStringLiteral(data));
          } else if (Array.isArray(data)) {
            items.push(createFromRawTransforms(data, ctx));
          }

          return factory.createArrayLiteralExpression(items);
        }),
      ),
    ),
  );
}
