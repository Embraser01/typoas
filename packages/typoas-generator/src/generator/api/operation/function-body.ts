import { Context } from '../../../context';
import { getQueryParams } from './query-params';
import { createRuntimeRefProperty, ExportedRef } from '../../utils/ref';
import { GlobalParameters } from './types';
import { isReferenceObject, OperationObject } from 'openapi3-ts/oas31';
import {
  Block,
  factory,
  NodeFlags,
  ObjectLiteralElementLike,
  Statement,
} from 'typescript';
import { createSchemaTransforms } from '../../utils/transformers';
import { getContentTypeSchema } from '../../utils/content-type';

export function createOperationBody(
  operation: OperationObject,
  path: string,
  method: string,
  ctx: Context,
  { baseParameters, defaultSecurityRequirements }: GlobalParameters,
): Block {
  const statements: Statement[] = [];

  const params = [...(operation.parameters || []), ...baseParameters];
  const queryParams = getQueryParams(params, ctx);

  const secRequirements = operation.security || defaultSecurityRequirements;
  const securities = new Set<string>(
    secRequirements?.flatMap((s) => Object.keys(s)),
  );

  // Create request
  statements.push(
    factory.createVariableStatement(
      undefined,
      factory.createVariableDeclarationList(
        [
          factory.createVariableDeclaration(
            'req',
            undefined,
            undefined,
            factory.createAwaitExpression(
              factory.createCallExpression(
                factory.createPropertyAccessExpression(
                  factory.createIdentifier('ctx'),
                  factory.createIdentifier('createRequest'),
                ),
                undefined,
                [
                  factory.createObjectLiteralExpression(
                    [
                      factory.createPropertyAssignment(
                        factory.createIdentifier('path'),
                        factory.createStringLiteral(path, true),
                      ),
                      factory.createShorthandPropertyAssignment(
                        factory.createIdentifier('params'),
                        undefined,
                      ),
                      factory.createPropertyAssignment(
                        factory.createIdentifier('method'),
                        factory.createPropertyAccessExpression(
                          createRuntimeRefProperty(ExportedRef.HttpMethod),
                          factory.createIdentifier(method.toUpperCase()),
                        ),
                      ),
                      ...(operation.requestBody
                        ? [
                            factory.createShorthandPropertyAssignment(
                              factory.createIdentifier('body'),
                              undefined,
                            ),
                          ]
                        : []),
                      ...(queryParams.length
                        ? [
                            factory.createPropertyAssignment(
                              factory.createIdentifier('queryParams'),
                              factory.createArrayLiteralExpression(
                                queryParams,
                                true,
                              ),
                            ),
                          ]
                        : []),
                      ...(securities.size
                        ? [
                            factory.createPropertyAssignment(
                              factory.createIdentifier('auth'),
                              factory.createArrayLiteralExpression(
                                [...securities.values()].map((s) =>
                                  factory.createStringLiteral(s),
                                ),
                                false,
                              ),
                            ),
                          ]
                        : []),
                    ],
                    true,
                  ),
                ],
              ),
            ),
          ),
        ],
        NodeFlags.Const,
      ),
    ),
  );

  // Send request
  statements.push(
    factory.createVariableStatement(
      undefined,
      factory.createVariableDeclarationList(
        [
          factory.createVariableDeclaration(
            factory.createIdentifier('res'),
            undefined,
            undefined,
            factory.createAwaitExpression(
              factory.createCallExpression(
                factory.createPropertyAccessExpression(
                  factory.createIdentifier('ctx'),
                  factory.createIdentifier('sendRequest'),
                ),
                undefined,
                [
                  factory.createIdentifier('req'),
                  ...(ctx.hasFetcherOptions()
                    ? [factory.createIdentifier('opts')]
                    : []),
                ],
              ),
            ),
          ),
        ],
        NodeFlags.Const,
      ),
    ),
  );

  // Handle response
  statements.push(
    factory.createReturnStatement(
      factory.createCallExpression(
        factory.createPropertyAccessExpression(
          factory.createIdentifier('ctx'),
          factory.createIdentifier('handleResponse'),
        ),
        undefined,
        [
          factory.createIdentifier('res'),
          factory.createObjectLiteralExpression(
            createOperationResponseHandlers(operation, ctx),
            true,
          ),
        ],
      ),
    ),
  );
  return factory.createBlock(statements, true);
}

export function createOperationResponseHandlers(
  operation: OperationObject,
  ctx: Context,
): ObjectLiteralElementLike[] {
  const responses = Object.entries(operation.responses || {});
  return responses
    .map(([code, res]) => {
      if (isReferenceObject(res)) {
        const ref = ctx.resolveReference('responses', res.$ref);
        if (!ref) {
          throw new Error(`$ref '${res.$ref}' not found`);
        }
        res = ref.spec;
      }
      const schema = getContentTypeSchema(res.content, ctx);
      if (!schema) {
        return null;
      }
      const transforms = createSchemaTransforms(schema, ctx);
      if (!transforms) {
        return null;
      }
      return factory.createPropertyAssignment(
        factory.createStringLiteral(code, true),
        factory.createObjectLiteralExpression([
          factory.createPropertyAssignment(
            factory.createIdentifier('transforms'),
            transforms,
          ),
        ]),
      );
    })
    .filter(Boolean) as ObjectLiteralElementLike[];
}
