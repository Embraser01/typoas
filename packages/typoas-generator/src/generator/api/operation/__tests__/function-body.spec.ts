import { describe, expect, it } from '@jest/globals';
import { createOperationResponseHandlers } from '../function-body';
import { getStringFromNode } from '../../../utils/ts-node';
import { Context } from '../../../../context';
import { factory } from 'typescript';

describe('create operation response handlers', () => {
  it('should create responses', () => {
    const node = factory.createObjectLiteralExpression(
      createOperationResponseHandlers(
        {
          responses: {
            '200': {
              content: {
                'application/json': {
                  schema: { type: 'string', format: 'date-time' },
                },
              },
            },
          },
        },
        new Context({ jsDoc: false }),
      ),
    );

    expect(getStringFromNode(node)).toMatchSnapshot();
  });
});
