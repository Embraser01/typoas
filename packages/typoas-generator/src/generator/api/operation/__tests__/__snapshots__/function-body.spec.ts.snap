// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create operation response handlers should create responses 1`] = `"{ '200': { transforms: { date: [[[r.TransformType.THIS]]] } } }"`;
